File: lib/python/openlinktoken/src/main/openlinktoken/tokentransformer/jwe_match_token_formatter.py
# SPDX-License-Identifier: MIT
"""
JWE Match Token Formatter for Open Link Token V1 format.
"""

import base64
import json
import time
from typing import Optional, Union

from jwcrypto import jwe, jwk

from openlinktoken.crypto_suite import CryptoSuite
from openlinktoken.tokentransformer.match_token_constants import (
    HEADER_KEY_ALGORITHM,
    HEADER_KEY_ENCRYPTION,
    HEADER_KEY_KEY_ID,
    HEADER_KEY_TYPE,
    PAYLOAD_KEY_HASH_ALGORITHM,
    PAYLOAD_KEY_ISSUED_AT,
    PAYLOAD_KEY_ISSUER,
    PAYLOAD_KEY_MAC_ALGORITHM,
    PAYLOAD_KEY_PPID,
    PAYLOAD_KEY_RING_ID,
    PAYLOAD_KEY_RULE_ID,
    TOKEN_TYPE,
    V1_TOKEN_PREFIX,
)
from openlinktoken.tokentransformer.token_transformer import TokenTransformer


class JweMatchTokenFormatter(TokenTransformer):
    """
    Formats tokens in the JWE-based match token format (olt.V1.<JWE>).

    This formatter wraps the privacy-protected identifier (PPID) in a
    self-contained JWE structure with all necessary metadata for versioning
    and cryptographic agility.

    See RFC 7516 - JSON Web Encryption (JWE)
    """

    def __init__(
        self,
        encryption_key: Union[str, bytes],
        ring_id: str,
        rule_id: str,
        issuer: Optional[str] = None,
        crypto_suite: CryptoSuite | None = None,
    ):
        """
        Initialize the JWE match token formatter.

        Accepts either a ``str`` (UTF-8 encoded; must encode to exactly 32 bytes) or
        raw ``bytes`` (must be exactly 32 bytes) for the encryption key.

        Args:
            encryption_key: The encryption key (must be exactly 32 bytes for AES-256).
            ring_id: The ring identifier for key management.
            rule_id: The token rule identifier (e.g., "T1", "T2", etc.).
            issuer: The issuer identifier (optional, defaults to "org.openlinktoken").
            crypto_suite: The suite whose digest and MAC metadata is embedded in the token.

        Raises:
            ValueError: If encryption_key, ring_id, or rule_id are invalid.
        """
        if isinstance(encryption_key, bytes):
            key_bytes = encryption_key
        else:
            key_bytes = encryption_key.encode("utf-8") if encryption_key else b""

        if not key_bytes or len(key_bytes) != 32:
            raise ValueError("Encryption key must be exactly 32 bytes (256 bits)")
        if not ring_id:
            raise ValueError("Ring ID must not be None or empty")
        if not rule_id:
            raise ValueError("Rule ID must not be None or empty")

        self.ring_id = ring_id
        self.rule_id = rule_id
        self.issuer = issuer if issuer else "org.openlinktoken"
        self.crypto_suite = crypto_suite or CryptoSuite.default()

        # Create JWK from the encryption key - needs to be base64url-encoded
        key_b64 = base64.urlsafe_b64encode(key_bytes).decode("utf-8").rstrip("=")
        self.jwk_key = jwk.JWK(kty="oct", k=key_b64)

    def transform(self, token: str) -> str:
        """
        Transform a token (PPID) into the JWE match token format.

        The input token should be the base64-encoded HMAC output from previous transformers.
        This method wraps it in a JWE structure with metadata and prepends the "olt.V1." prefix.

        Args:
            token: The privacy-protected identifier (PPID) to wrap in JWE format

        Returns:
            The formatted match token: olt.V1.<JWE compact serialization>

        Raises:
            Exception: If JWE encryption or serialization fails
        """
        if not token:
            # Return as-is for blank tokens
            return token

        try:
            # Build the JWE payload with metadata
            payload = {
                PAYLOAD_KEY_RULE_ID: self.rule_id,
                PAYLOAD_KEY_HASH_ALGORITHM: self.crypto_suite.token_digest_algorithm,
                PAYLOAD_KEY_MAC_ALGORITHM: self.crypto_suite.token_mac_algorithm,
                PAYLOAD_KEY_PPID: [token],  # PPID as an array (single element for hash-based tokens)
                PAYLOAD_KEY_RING_ID: self.ring_id,
                PAYLOAD_KEY_ISSUER: self.issuer,
                PAYLOAD_KEY_ISSUED_AT: int(time.time()),
            }

            # Create JWE header with algorithm and encryption method
            protected_header = {
                HEADER_KEY_ALGORITHM: "dir",  # Direct encryption (key used directly)
                HEADER_KEY_ENCRYPTION: "A256GCM",  # AES-256-GCM encryption
                HEADER_KEY_TYPE: TOKEN_TYPE,
                HEADER_KEY_KEY_ID: self.ring_id,
            }

            # Create JWE object
            jwe_token = jwe.JWE(
                plaintext=json.dumps(payload).encode("utf-8"), recipient=self.jwk_key, protected=protected_header
            )

            # Serialize to compact form and prepend the olt.V1. prefix
            jwe_compact = jwe_token.serialize(compact=True)
            return V1_TOKEN_PREFIX + jwe_compact

        except Exception as e:
            raise Exception(f"JWE token generation failed for rule {self.rule_id}: {str(e)}")

File: lib/java/openlinktoken/src/main/java/org/openlinktoken/tokentransformer/JweMatchTokenFormatter.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.tokentransformer;

import com.nimbusds.jose.EncryptionMethod;
import com.nimbusds.jose.JOSEException;
import com.nimbusds.jose.JWEAlgorithm;
import com.nimbusds.jose.JWEHeader;
import com.nimbusds.jose.JOSEObjectType;
import com.nimbusds.jose.JWEObject;
import com.nimbusds.jose.Payload;
import com.nimbusds.jose.crypto.DirectEncrypter;
import com.nimbusds.jose.jwk.OctetSequenceKey;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import org.openlinktoken.crypto.CryptoSuite;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Formats tokens in the JWE-based match token format (olt.V1.&lt;JWE&gt;).
 * <p>
 * This transformer wraps the privacy-protected identifier (PPID) in a
 * self-contained JWE structure with all necessary metadata for versioning
 * and cryptographic agility.
 *
 * @see <a href="https://datatracker.ietf.org/doc/html/rfc7516">RFC 7516 - JSON Web Encryption (JWE)</a>
 */
public class JweMatchTokenFormatter implements TokenTransformer {
    private static final long serialVersionUID = 1L;
    private static final Logger logger = LoggerFactory.getLogger(JweMatchTokenFormatter.class);

    private final String ringId;
    private final String ruleId;
    private final String issuer;
    private final byte[] encryptionKey;
    private final CryptoSuite cryptoSuite;
    private transient DirectEncrypter encrypter;

    /**
     * Initializes the JWE match token formatter.
     *
     * @param encryptionKey the encryption key (must be 32 bytes for AES-256)
     * @param ringId the ring identifier for key management
     * @param ruleId the token rule identifier (e.g., "T1", "T2", etc.)
     * @param issuer the issuer identifier (optional, defaults to "org.openlinktoken")
     * @throws JOSEException if the encrypter cannot be initialized
     */
    public JweMatchTokenFormatter(String encryptionKey, String ringId, String ruleId, String issuer)
            throws JOSEException {
        this(encryptionKey == null ? null : encryptionKey.getBytes(StandardCharsets.UTF_8), ringId, ruleId, issuer,
                CryptoSuite.defaultSuite());
    }

    /**
     * Initializes the JWE match token formatter using raw encryption key bytes.
     *
     * @param encryptionKey the raw encryption key (must be exactly 32 bytes for AES-256)
     * @param ringId the ring identifier for key management
     * @param ruleId the token rule identifier (e.g., "T1", "T2", etc.)
     * @param issuer the issuer identifier (optional, defaults to "org.openlinktoken")
     * @throws JOSEException if the encrypter cannot be initialized
     */
    public JweMatchTokenFormatter(byte[] encryptionKey, String ringId, String ruleId, String issuer)
            throws JOSEException {
        this(encryptionKey, ringId, ruleId, issuer, CryptoSuite.defaultSuite());
    }

    /**
     * Initializes the JWE formatter with an explicit crypto suite.
     *
     * @param encryptionKey the raw encryption key
     * @param ringId the key ring identifier
     * @param ruleId the token rule identifier
     * @param issuer the token issuer
     * @param cryptoSuite the suite whose token metadata is embedded
     * @throws JOSEException if the encrypter cannot be initialized
     */
    public JweMatchTokenFormatter(
            byte[] encryptionKey,
            String ringId,
            String ruleId,
            String issuer,
            CryptoSuite cryptoSuite) throws JOSEException {
        byte[] keyBytes = validateEncryptionKey(encryptionKey);
        if (ringId == null || ringId.isEmpty()) {
            throw new IllegalArgumentException("Ring ID must not be null or empty");
        }
        if (ruleId == null || ruleId.isEmpty()) {
            throw new IllegalArgumentException("Rule ID must not be null or empty");
        }

        this.ringId = ringId;
        this.ruleId = ruleId;
        this.issuer = (issuer != null && !issuer.isEmpty()) ? issuer : "org.openlinktoken";
        this.encryptionKey = keyBytes;
        this.cryptoSuite = cryptoSuite;
        this.encrypter = createEncrypter(this.encryptionKey);
    }

    private void writeObject(ObjectOutputStream oos) throws IOException {
        oos.defaultWriteObject();
    }

    private void readObject(ObjectInputStream ois) throws IOException, ClassNotFoundException {
        ois.defaultReadObject();
        try {
            this.encrypter = createEncrypter(this.encryptionKey);
        } catch (JOSEException e) {
            throw new IOException("Failed to reconstruct JWE encrypter", e);
        }
    }

    private static byte[] validateEncryptionKey(byte[] encryptionKey) {
        if (encryptionKey == null || encryptionKey.length != 32) {
            throw new IllegalArgumentException("Encryption key must be exactly 32 bytes (256 bits)");
        }
        return Arrays.copyOf(encryptionKey, encryptionKey.length);
    }

    private static DirectEncrypter createEncrypter(byte[] encryptionKey) throws JOSEException {
        OctetSequenceKey jwk = new OctetSequenceKey.Builder(encryptionKey).build();
        return new DirectEncrypter(jwk);
    }

    /**
     * Transforms a token (PPID) into the JWE match token format.
     * <p>
     * The input token should be the base64-encoded HMAC output from previous transformers.
     * This method wraps it in a JWE structure with metadata and prepends the "olt.V1." prefix.
     *
     * @param token the privacy-protected identifier (PPID) to wrap in JWE format
     * @return the formatted match token: olt.V1.&lt;JWE compact serialization&gt;
     * @throws Exception if JWE encryption or serialization fails
     */
    @Override
    public String transform(String token) throws Exception {
        if (token == null || token.isBlank()) {
            logger.warn("Received null or empty token for rule {}", ruleId);
            return token; // Return as-is for blank tokens
        }

        try {
            // Build the JWE payload with metadata using a Map
            Map<String, Object> payload = new LinkedHashMap<>();
            payload.put(MatchTokenConstants.PAYLOAD_KEY_RULE_ID, ruleId);
            payload.put(MatchTokenConstants.PAYLOAD_KEY_HASH_ALGORITHM, cryptoSuite.getTokenDigestAlgorithm());
            payload.put(MatchTokenConstants.PAYLOAD_KEY_MAC_ALGORITHM, cryptoSuite.getTokenMacAlgorithm());
            payload.put(MatchTokenConstants.PAYLOAD_KEY_PPID, Collections.singletonList(token));
            payload.put(MatchTokenConstants.PAYLOAD_KEY_RING_ID, ringId);
            payload.put(MatchTokenConstants.PAYLOAD_KEY_ISSUER, issuer);
            payload.put(MatchTokenConstants.PAYLOAD_KEY_ISSUED_AT, Instant.now().getEpochSecond());

            // Create JWE header with algorithm and encryption method
            JWEHeader header = new JWEHeader.Builder(JWEAlgorithm.DIR, EncryptionMethod.A256GCM)
                    .type(new JOSEObjectType(MatchTokenConstants.TOKEN_TYPE))
                    .keyID(ringId)
                    .build();

            // Create JWE object with the payload (Payload accepts Map and converts to JSON)
            JWEObject jweObject = new JWEObject(header, new Payload(payload));

            // Encrypt the JWE object
            jweObject.encrypt(encrypter);

            // Serialize to compact form and prepend the olt.V1. prefix
            String jweCompact = jweObject.serialize();
            return MatchTokenConstants.V1_TOKEN_PREFIX + jweCompact;

        } catch (JOSEException e) {
            logger.error("Failed to create JWE token for rule {}: {}", ruleId, e.getMessage());
            throw new Exception("JWE token generation failed", e);
        }
    }
}
